### PR TITLE
Integer-valued VectorSpace dimensions

### PR DESCRIPTION
**Pull request by peoe**
This PR proposes #2062.

I'm a bit torn about his implementation because most `VectorSpace`s use slightly different methods to compute their dimension. If you have ideas on how to unify these, please let me know.

### CODE REVIEW

**Comment by peoe** [comment]
@pmli : Does this better suit the "don't hide issues" mentatlity? This way, the setter handles type checking and I also got rid of additional getter implementations.

File: src/pymor/vectorarrays/interface.py [hunk 1]
```diff
@@ -970,6 +970,16 @@ def from_numpy(self, data, ensure_copy=False):
         """
         raise NotImplementedError
 
+    @property
+    def dim(self):
+        return self._dim
+
+    @dim.setter
+    def dim(self, dim):
+        if not isinstance(dim, int):
```
Comment (pmli) [inline src/pymor/vectorarrays/interface.py:979]: This fails on NumPy integers:

```python
>>> import numpy as np
>>> a = np.int32(5)
>>> a
5
>>> isinstance(a, int)
False
```

Using `numbers.Integral` makes it work:

```python
>>> from numbers import Integral
>>> isinstance(a, Integral)
True
```

Still, that doesn't work with 0-dimensional NumPy arrays, where checking the dimension and dtype seems to be necessary:

```python
>>> b = np.array(5)
>>> b
array(5)
>>> print(b)
5
>>> isinstance(b, Integral)
False
>>> b.ndim == 0 and np.issubdtype(b.dtype, np.integer)
True
```

I'm not sure how far we need to go. Maybe it's ok to just limit to the `Integral` check and see if issues come up.

**Review by pmli — COMMENTED** [review]
The trick with the setter looks very good. I added a comment about type-checking.

File: src/pymor/vectorarrays/interface.py [hunk 2]
```diff
@@ -970,6 +970,16 @@ def from_numpy(self, data, ensure_copy=False):
...
+            raise TypeError(f'unsupported dtype={type(dim)}')
+        self._dim = dim
```
Comment (sdrave) [inline src/pymor/vectorarrays/interface.py:981]: I'm not quite sure, but doesn't this make the dimension of a `VectorSpace` mutable? This should not be allowed ..

File: src/pymor/bindings/fenics.py [hunk 3]
```diff
@@ -138,10 +138,7 @@ class FenicsVectorSpace(ComplexifiedListVectorSpace):
 
     def __init__(self, V, id='STATE'):
         self.__auto_init(locals())
-
-    @property
-    def dim(self):
-        return df.Function(self.V).vector().size()
+        self.dim = df.Function(self.V).vector().size()
```
Comment (sdrave) [inline src/pymor/bindings/fenics.py:141]: This means that creating a new `FenicsVectorSpace` always entails creating a new `Function` object and, thus, allocating memory for a DOF vector that is never used. I generally assume that creating a `VectorSpace` is cheap.
Note that there are very few places in pyMOR where the dimension of a `VectorSpace` actually is relevant.

(The current code is not optimal either, of course. The dimension should be cached.)

**Review by sdrave — CHANGES_REQUESTED** [review]
Apart from my comments, I feel that this approach is unnecessarily fancy. I'd just check the argument in `NumpyVectorSpace.__init__`. Most spaces derive there dimension from other data anyway.

**Comment by sdrave** [comment]
Closing in favor of #2081.

**Pull request closed without merging**